rholang/src/main/scala/coop/rchain/rholang/interpreter/Interpreter.scala

```diff
@@ -60,30 +68,55 @@ object Interpreter {
         sortedPar <- Sortable[Par].sortMatch(par)
       } yield sortedPar.term
 
-    def execute(runtime: Runtime[F], reader: Reader): F[Runtime[F]] =
-      for {
-        par    <- buildNormalizedTerm(reader)
-        errors <- evaluate(runtime, par).map(_.errors)
-        result <- if (errors.isEmpty) F.pure(runtime)
-                 else F.raiseError(new RuntimeException(mkErrorMsg(errors)))
-      } yield result
+    def evaluatePar(runtime: Runtime[F], par: Par): F[EvaluateResult] = {
+      val initialPhlo = Cost.Max
+      evaluatePar(runtime, par, initialPhlo)
+    }
+
+    def evaluatePar(runtime: Runtime[F], par: Par, initialPhlo: Cost): F[EvaluateResult] = {
+      val term = PrettyPrinter().buildString(par)
+      evaluate(runtime, term, initialPhlo)
+    }
 
-    def evaluate(runtime: Runtime[F], par: Par, initialPhlo: Cost): F[EvaluateResult] = {
+    def evaluate(runtime: Runtime[F], term: String): F[EvaluateResult] =
+      evaluate(runtime, term, Cost.Max)
+
+    def evaluate(runtime: Runtime[F], term: String, initialPhlo: Cost): F[EvaluateResult] = {
       implicit val rand: Blake2b512Random = Blake2b512Random(128)
       for {
-        checkpoint    <- runtime.space.createCheckpoint()
-        _             <- runtime.reducer.setPhlo(initialPhlo)
-        _             <- runtime.reducer.inj(par)(rand)
-        errors        <- runtime.readAndClearErrorVector()
-        remainingPhlo <- runtime.reducer.phlo
-        cost          = initialPhlo - remainingPhlo
-        _             <- if (errors.nonEmpty) runtime.space.reset(checkpoint.root) else F.unit
-      } yield EvaluateResult(cost, errors)
+        checkpoint <- runtime.space.createCheckpoint()
+        res        <- injAttempt(runtime.reducer, runtime.errorLog, term, initialPhlo)
```

Comment:
I like how you moved 'setPhlo' inside the injAttempt, and are returning EvaluateResult from it. Those are deduplications long due :+1: